ext/node/ops/ipc.rs

```diff
@@ -134,106 +372,161 @@ mod impl_ {
       })
     }
 
-    #[cfg(unix)]
-    #[cfg(test)]
-    fn from_stream(stream: UnixStream) -> Self {
+    #[cfg(all(unix, test))]
+    fn from_stream(stream: UnixStream, ref_tracker: IpcRefTracker) -> Self {
       let (read_half, write_half) = stream.into_split();
       Self {
         read_half: AsyncRefCell::new(IpcJsonStream::new(read_half)),
         write_half: AsyncRefCell::new(write_half),
         cancel: Default::default(),
+        queued_bytes: Default::default(),
+        ref_tracker,
       }
     }
 
-    #[cfg(windows)]
-    #[cfg(test)]
-    fn from_stream(pipe: NamedPipeClient) -> Self {
+    #[cfg(all(windows, test))]
+    fn from_stream(
+      pipe: NamedPipeClient,
+      ref_tracker: ExternalOpsTracker,
+    ) -> Self {
       let (read_half, write_half) = tokio::io::split(pipe);
       Self {
         read_half: AsyncRefCell::new(IpcJsonStream::new(read_half)),
         write_half: AsyncRefCell::new(write_half),
         cancel: Default::default(),
+        queued_bytes: Default::default(),
+        refed: AtomicBool::new(false),
+        ref_tracker,
       }
     }
 
-    async fn write_msg(
+    /// writes _newline terminated_ JSON message to the IPC pipe.
+    async fn write_msg_bytes(
       self: Rc<Self>,
-      msg: serde_json::Value,
+      msg: &[u8],
+    ) -> Result<(), AnyError> {
+      let mut write_half =
+        RcRef::map(self, |r| &r.write_half).borrow_mut().await;
+      write_half.write_all(msg).await?;
+      Ok(())
+    }
+
+    #[cfg_attr(not(test), allow(dead_code))]
+    async fn write_msg<T: serde::Serialize>(
+      self: Rc<Self>,
+      msg: &T,
     ) -> Result<(), AnyError> {
       let mut write_half =
         RcRef::map(self, |r| &r.write_half).borrow_mut().await;
       // Perf note: We do not benefit from writev here because
       // we are always allocating a buffer for serialization anyways.
-      let mut buf = Vec::new();
-      serde_json::to_writer(&mut buf, &msg)?;
+      let mut buf = serde_json::to_vec(msg)?;
       buf.push(b'\n');
       write_half.write_all(&buf).await?;
       Ok(())
     }
   }
 
-  #[inline]
-  fn memchr(needle: u8, haystack: &[u8]) -> Option<usize> {
-    #[cfg(all(target_os = "macos", target_arch = "aarch64"))]
-    // Safety: haystack of valid length. neon_memchr can handle unaligned
-    // data.
-    return unsafe { neon::neon_memchr(haystack, needle, haystack.len()) };
-
-    #[cfg(not(all(target_os = "macos", target_arch = "aarch64")))]
-    return haystack.iter().position(|&b| b == needle);
-  }
-
   // Initial capacity of the buffered reader and the JSON backing buffer.
   //
   // This is a tradeoff between memory usage and performance on large messages.
   //
   // 64kb has been chosen after benchmarking 64 to 66536 << 6 - 1 bytes per message.
   const INITIAL_CAPACITY: usize = 1024 * 64;
 
+  // A buffer for reading from the IPC pipe.
+  // Similar to the internal buffer of `tokio::io::BufReader`.
+  struct ReadBuffer {
```

Comment:
Why not use `BufReader` instead? What's the difference?

Reply:
The original version of this code used a `BufReader`, but it had an optimization where if the entire JSON message fits in the buffer, it would just parse it in place (without copying to an intermediate buffer). The catch was that `simdjson` mutates buffers in place during parsing, so you need a mutable reference to the buffer.

`BufReader` only gives you immutable access, so the old code was just transmuting the buffer from `&[u8]` to `&mut [u8]`, which is UB.

To keep that optimization but also keep this safe, I just extracted out the bits we need (which it turns out is not much).

Reply:
Sounds good, can you add a comment about need for mutable reference and simdjson?